wallet-new/src/Cardano/Wallet/Types/UtxoStatistics.hs

```diff
@@ -0,0 +1,219 @@
+                     deriveSafeBuildable)
+
+
+--  Utxo statistics for the wallet.
+--  Histogram is composed of bars that represent the bucket. The bucket is tagged by upper bound of a given bucket.
+--  The bar value corresponds to the number of stakes
+--  In the future the bar value could be different things:
+--  (a) sum of stakes in a bucket
+--  (b) avg or std of stake in a bucket
+--  (c) topN buckets
+--  to name a few
+data HistogramBar = HistogramBarCount
+    { bucketName       :: !Text
+    , bucketUpperBound :: !Word64
+    } deriving (Show, Eq, Ord, Generic)
+
+--  Buckets boundaries can be constructed in different way
+data BoundType = Log10 | Haphazard
+
+generateBounds :: BoundType -> NonEmpty Word64
+generateBounds bType =
+    let (^!) :: Word64 -> Word64 -> Word64
+        (^!) = (^)
+    in case bType of
+        Log10 -> NL.fromList $ ( map (\toPower -> 10 ^! toPower) [1..16] ) ++ [45 * (10 ^! 15)]
+        Haphazard -> NL.fromList [10, 100, 1000, 10000]
+
+instance Arbitrary HistogramBar where
+    arbitrary = do
+        possiblenames <- elements $ map show (NL.toList $ generateBounds Log10)
+        bound <- arbitrary
+        pure (HistogramBarCount possiblenames bound)
+
+
+deriveSafeBuildable ''HistogramBar
+instance BuildableSafeGen HistogramBar where
+    buildSafeGen _ HistogramBarCount{..} =
+        bprint ("{"
+                %" name="%build
+                %" upperBound="%build
+                %" }")
+        bucketName
+        bucketUpperBound
+
+
+data UtxoStatistics = UtxoStatistics
+    { theHistogram :: ![HistogramBar]
+    , theAllStakes :: !Word64
+    } deriving (Show, Generic, Ord)
+
+toMap :: [HistogramBar] -> Map Text Word64
+toMap = Map.fromList . map (\(HistogramBarCount key val) -> (key,val))
+
+instance Eq UtxoStatistics where
+    (UtxoStatistics h s) == (UtxoStatistics h' s') = s == s' && toMap h == toMap h'
+
+instance ToJSON UtxoStatistics where
+    toJSON (UtxoStatistics bars allStakes) =
+        let histogramObject = Object . HMS.fromList . map extractBarKey
+            extractBarKey (HistogramBarCount bound stake) = bound .= stake
+        in object [ "histogram" .= histogramObject bars
+                  , "allStakes" .= allStakes ]
+
+instance FromJSON UtxoStatistics where
+    parseJSON = withObject "UtxoStatistics" $ \o -> do
+        histo <- o .: "histogram" :: Parser (HashMap Text Word64)
+        stakes <- o .: "allStakes"
+        case validateUtxoStatistics histo stakes of
+            Right (histogram, allStakes) -> do
+                let constructHistogram = uncurry HistogramBarCount
+                let histoBars = map constructHistogram $ HMS.toList histogram
+                pure $ UtxoStatistics histoBars allStakes
+            Left err -> fail $ "Failed to parse UtxoStatistics: " <> show err
+
+data UtxoStatisticsError
+    = ErrHistogramEmpty
+    | ErrHistogramNamesInvalid
+    | ErrHistogramUpperBoundsNegative
+    | ErrAllStakesNegative
```

Comment:
What if the total stake is lower than the minimum amount that can be represented with all the buckets :stuck_out_tongue: ?

Reply:
addressed